src/modm/math/geometry/vector.hpp

```diff
@@ -11,141 +12,552 @@
+#pragma once
 
-#ifndef MODM_VECTOR_HPP
-#define MODM_VECTOR_HPP
-
+#include <array>
 #include <cmath>
-#include <stdint.h>
+#include <concepts>
+#include <functional>
+#include <numeric>
+#include <ranges>
 
-#include <modm/math/matrix.hpp>
 #include <modm/math/utils/arithmetic_traits.hpp>
+#include <modm/math/utils/integer_traits.hpp>
+#include <modm/io/iostream.hpp>
+
+#include "../matrix.hpp"
+#include "angle.hpp"
 
 namespace modm
 {
-	// forward declaration
-	template<typename T, uint8_t W, uint8_t H> class Matrix;
+
+/**
+ * \brief	Class for handling common vector operations
+ *
+ * Arithmetic Operations:
+ * \code
+ *	+ : plus translation
+*	- : minus translation
+*	* : scalar multiplication / dot product
+*	/ : scalar division
+*	^ : cross product / determinant
+*	~ : perpendicular / orthogonal
+* \endcode
+*
+* \author	Fabian Greif
+* \author	Niklas Hauser
+* \author	Thomas Sommer
+* \ingroup	modm_math_geometry
+*/
+template<class T, std::size_t N>
+requires (N > 0)
+class Vector : public std::array<T, N>
+{
+	using WideType = std::conditional<std::is_floating_point_v<T>, T, modm::WideType<T>>::type;
+	using FloatType = std::conditional<std::is_floating_point_v<T>, T, float>::type;
+
+	template<std::size_t K>
+	void assign(){};
+
+	template<std::size_t K, typename U, class... Args>
+	requires std::convertible_to<U, T>
+	void assign(U v, Args... args)
+	{
+		if constexpr(std::integral<T> && std::floating_point<U>)
+			*(this->begin() + K) = std::round(v);
+		else
+			*(this->begin() + K) = v;
+
+		assign<K + 1>(args...);
+	}
+
+ 	template<std::size_t K, typename U, std::size_t M, class... Args>
+	// TODO One variadic method for modm::Vector and std::array
+	// requires std::derived_from<..., std::array>
+	void assign(const Vector<U, M> &vec, Args... args)
+	{
+		if constexpr(std::integral<T> && std::floating_point<U>)
+			std::transform(vec.begin(), vec.end(), this->begin() + K, [](U v) { return std::round(v); });
+		else
+			std::copy(vec.begin(), vec.end(), this->begin() + K);
+
+		assign<K + M>(args...);
+	}
+
+	template<std::size_t K, typename U, std::size_t M, class... Args>
+	void assign(const std::array<U, M> &arr, Args... args)
+	{
+		if constexpr(std::integral<T> && std::floating_point<U>)
+			std::transform(arr.begin(), arr.end(), this->begin() + K, [](U v) { return std::round(v); });
+		else
+			std::copy(arr.begin(), arr.end(), this->begin() + K);
+
+		assign<K + M>(args...);
+	}
+
+public:
+	constexpr Vector()
+	{ this->fill(0); }
 
 	/**
-	 * \brief	Class for handling common point operations
+	 * @brief 			Multi purpose constructor.
 	 *
-	 * Basic data type of all geometric operations. Used to represent vectors
-	 * as well as particular points in the coordinate system.
-	 *
-	 * \section	point_vector	Point vs. vector
+	 * @param args		scalars, Vectors and Arrays in your desired order
+	 */
+	template <class... Args>
+	constexpr Vector(Args... args)
+	{ assign<0>(args...); }
+
```

Comment:
This constructor replaces the original constructor-blocks, satisfies even more cases and is easily extended.